### PR TITLE
DOC: Add PyArray_FailUnlessWriteable to the online C-API docs. 

### LINKED ISSUE #20623: DOC: 'PyArray_FailUnlessWriteable' is not included in the C API documentation
### Issue with current documentation:

The function `PyArray_FailUnlessWriteable` (added in https://github.com/numpy/numpy/pull/280) does not show up in the [C API documentation](https://numpy.org/devdocs/reference/c-api/index.html).  The function is defined in
https://github.com/numpy/numpy/blob/bb95cf015165bcf85f99ba0d28a7a6e727ca955b/numpy/core/src/multiarray/arrayobject.c#L622-L645

### PR DESCRIPTION
**Pull request by WarrenWeckesser**
The text was copy-edited a bit, and the comments in the C file
were updated to match the documentation in the .rst file.

Closes gh-20623.

### CODE REVIEW

File: doc/source/release/1.7.0-notes.rst [hunk 1]
```diff
@@ -162,7 +162,7 @@ Added experimental support for the AArch64 architecture.
 C API
 -----
 
-New function ``PyArray_RequireWriteable`` provides a consistent interface
+New function ``PyArray_FailUnlessWriteable`` provides a consistent interface
```
Comment (mattip) [inline doc/source/release/1.7.0-notes.rst:165]: For other reviewers who are confused like I was, this is really a typo in the 1.7.0 release note. PR #280, which was part of 1.7, renamed the function before its first release, and somehow the release note was not updated.

**Comment by mattip** [comment]
@seiko2plus I am trying to get the doxygen integration to work here. I made these changes to this PR but it is giving me this warning. I see the function in the index.xml of doxygen. Can you figure out what is going on?
```
doc/source/reference/c-api/array.rst:1626: WARNING: doxygenfunction: \
    Cannot find function "PyArray_FailUnlessWriteable" in \
    doxygen xml output for project "numpy" \
    from directory: ../build/doxygen/xml
```

<details><summary>Diff to this PR</summary>

```diff
diff --git a/doc/source/doxyfile b/doc/source/doxyfile
index ea45b9578..08a2de54b 100644
--- a/doc/source/doxyfile
+++ b/doc/source/doxyfile
@@ -118,7 +118,7 @@ WARN_LOGFILE           =
 #---------------------------------------------------------------------------
 INPUT                  =
 INPUT_ENCODING         = UTF-8
-FILE_PATTERNS          = *.h, *.hpp
+FILE_PATTERNS          = *.h, *.hpp, *.c, *.cpp
 RECURSIVE              = YES
 EXCLUDE                =
 EXCLUDE_SYMLINKS       = NO
diff --git a/doc/source/reference/c-api/array.rst b/doc/source/reference/c-api/array.rst
index d2e873802..c7b8f1628 100644
--- a/doc/source/reference/c-api/array.rst
+++ b/doc/source/reference/c-api/array.rst
@@ -1623,17 +1623,7 @@ For all of these macros *arr* must be an instance of a (subclass of)
     calculations in NumPy that rely on the state of these flags do not
     repeat the calculation to update them.
 
-.. c:function:: int PyArray_FailUnlessWriteable(PyArrayObject *obj, const char *name)
-
-    This function does nothing and returns 0 if *obj* is writeable.
-    It raises an exception and returns -1 if *obj* is not writeable.
-    It may also do other house-keeping, such as issuing warnings on
-    arrays which are transitioning to become views. Always call this
-    function at some point before writing to an array.
-
-    *name* is a name for the array, used to give better error messages.
-    It can be something like "assignment destination", "output array",
-    or even just "array".
+.. doxygenfunction:: PyArray_FailUnlessWriteable
 
 Array method alternative API
 ----------------------------
diff --git a/numpy/core/src/multiarray/arrayobject.c b/numpy/core/src/multiarray/arrayobject.c
index a96ab31b9..a513f3fb5 100644
--- a/numpy/core/src/multiarray/arrayobject.c
+++ b/numpy/core/src/multiarray/arrayobject.c
@@ -619,7 +619,7 @@ array_might_be_written(PyArrayObject *obj)
     return 0;
 }
 
-/*NUMPY_API
+/**
  *
  *  This function does nothing and returns 0 if *obj* is writeable.
  *  It raises an exception and returns -1 if *obj* is not writeable.
@@ -631,6 +631,7 @@ array_might_be_written(PyArrayObject *obj)
  *  It can be something like "assignment destination", "output array",
  *  or even just "array".
  */
+/* NUMPY_API */
 NPY_NO_EXPORT int
 PyArray_FailUnlessWriteable(PyArrayObject *obj, const char *name)
 {
```

If we cannot figure it out quickly, I think we should just merge as-is.

**Comment by charris** [comment]
Thanks Warren. @mattip We can figure out the doxygen problem later.

**Comment by seiko2plus** [comment]
@mattip, sorry for the delayed response, you have to [feed doxygen](https://numpy.org/devdocs/dev/howto-docs.html#feeding-doxygen) with the desired header through the sub-config files. just create a new sub-config file inside the path `numpy/core/src/multiarray/.doxyfile` contains:
```bash
INPUT += @CUR_DIR/FILE_NAME
```

**Pull request merged**